### ISSUE TITLE
Improve the Range for Type Helpers like `ValueOf`

### ISSUE THREAD

**Issue by ajitjha393** [ts-essentials/ts-essentials#333]
### Definition

```ts
 type ValueOf<T> = T[keyof T] 
```
As per my understanding, the Domain of the `valueOf` type helper since being generic can be any Type Parameter (T)
Hence the Range for this Mapping should also be defined to cover it.

As you can see for example 2 lists out all the properties of the string that might not be sound given the Domain as a string.

```ts 
const obj = {
  id: "123e4567-e89b-12d3-a456-426655440000",
  name: "Test object",
  timestamp: 1548768231486,
};

type ObjectValueType = ValueOf<typeof obj>;
//   ^? string | number

const obj2 = ""
type Ex2 = ValueOf<typeof obj2> // list out all props of string domain set
// ^? string | number | (() => IterableIterator<string>) | (() => string) | ...

```

Cases like this are not handled in the implementation and it might lead to some strange results and affect the `ValueOf` type being used as well...

So what should be the Output for such Inputs which are in the Domain of Primitives?

One way can be to make it act like an Identity function. 
That is if the `ValueOf` is being passed a Primitive we can simply return that type.
Hence the Mapping will be `(D -> R)` and would cover all bases.

Proposed Solution - 

```ts
type Primitive = string | number | boolean | bigint | symbol | undefined | null;

type ValueOf<T> = T extends Primitive ? T: T[keyof T]

let obj1 = {name: "bishwajit", age: 22}

type Ex1 = ValueOf<typeof obj1>
    // ^? string | number

let obj2 = "hey"

type Ex2 = ValueOf<typeof obj2>
  // ^? string

const obj3 = "hey"

type Ex3 = ValueOf<typeof obj3>
   // ^? "hey"
 

```
 
 [Code Playground](https://www.typescriptlang.org/play?#code/C4TwDgpgBACgTgSwLYOAgbtAvFAzsRAOwHMoAfKQgVyQCMI5ypaB7FgGwgENCnaFiCQsCa4QdDkyqEAJhABmQiDKbV27ANwBYAFC7QkKADUu7KhADy8gDwAVAHxQctqBAAewCLNyxEKNJhQAPxQtgBcoQDaANYQICzyoQC6urqcIiy0AFYAjE5QAN6EXEgQEQBE-LgAFgDuXFmo5QA0UFzEZVAATF0Avqk6BtAAom55OCZmljZDCVCZufa6UCtQAPRrUAB6QQPp89ld+eXVceUDQ1CjRxOm5lbWs4kLXUs6Kxvbu3o6AMYshHwByyAGZjqcQOcfpdRmDblMHk9gSC3qtPjtlkA)
 
 
 Let me know what are your thoughts @Beraliv .
 If looks good I can raise a PR

**Comment by Beraliv** [comment]
Hey Bishwajit @ajitjha393!

Thank you for your issue!

Yeah, there's an area of improvements here as it's not properly defined / tested and clarified from what I see in documentation:

## Implementation

```ts
/** Easily extract the type of a given object's values */
export type ValueOf<T> = T[keyof T];
```

From what I see here, it means that it's expected to pass only objects, everything else is up to you if you know what you're doing.

But I wonder why we don't have generic constrain like `type ValueOf<T extends Record<string, unknown>>`? Good question, maybe we don't need it as we allow to do other things and it's time to define a domain as you've explained

## Testing

Completely missed and that just explains that we don't suggest anything and it's good to clarify it

## Documentation (README)

Same story as in Implementation – we expect to pass only object and have this example https://github.com/ts-essentials/ts-essentials#ValueOf-type:

```ts
const obj = {
  id: "123e4567-e89b-12d3-a456-426655440000",
  name: "Test object",
  timestamp: 1548768231486,
};

type ObjectValueType = ValueOf<typeof obj>;
//   ^? string | number
```

So I see your point and I agree that it can be improved

So I'm happy to do the following

## Plan

0. Cover existing implementation with tests (primitives, objects, arrays, you can check other examples to understand that approach, I can obviously can help you with it)

1. For [Primitives](https://github.com/ts-essentials/ts-essentials/blob/master/lib/types.ts#L4) we can map it to the same value explaining it as we cannot really do anything with primitives

2. For objects we leave the same behaviour (checking that it's not broken with tests written at Step 0)

Open questions:

1. Define what we want to see for arrays and make it deterministic (e.g. exclude unnecessary information – https://tsplay.dev/mx87Xm)

2. Cover more edge cases in tests (e.g. array | undefined, object | primitive) – up to you to define it, I'm just happy to cover as much as possible, e.g. https://tsplay.dev/W4ngBN

Let's quickly clarify open questions and feel free to open PR if you have free time.

I'm happy to assist you when you start, just ping me and I will help you through it

**Comment by ajitjha393** [comment]
Hey @Beraliv,
Thanks for the acknowledgment!
I am totally on board with the above-mentioned plan for this. It will cover all our edge cases as well.

We need to plan it out for all the values and combinations with which `ValueOf` can be instantiated
- **primitives**
- **object**
- **array**
- **function**
- **combination of above types**

I think if we defined the behavior and expected Range for the above 4 domains, then a combination of the above types will be handled as per the `Distributed Law of Conditional Types`

```ts
type ValueOf<T> = 
    | T extends Primitive? T 
    : T extends any[]? T[number] 
    : T extends (...args: any[]) => any 
    ? ReturnType<T> : T[keyof T];
```
Code Playground - [https://tsplay.dev/WKOQGW](https://tsplay.dev/WKOQGW)

We can cover all these scenarios by writing tests for each of them as per the implementation.

**Comment by Beraliv** [comment]
Released in v9.3.1 ✅
